lib/scatter_swap/hasher.rb

```diff
@@ -1,81 +1,136 @@
+      if (!@_swapper_maps || !@_swapper_maps[index])
+        @_swapper_maps ||= []
+        index_swapper_map = []
+
+        array = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+        10.times do |i|
+          index_swapper_map << array.rotate!(index + i ^ spin).pop
+        end
+
+        @_swapper_maps[index] = index_swapper_map
       end
+
+      return @_swapper_maps[index]
     end
 
     # Using a unique map for each of the ten places,
     # we swap out one number for another
-    def swap
-      @working_array = @working_array.collect.with_index do |digit, index|
+    def swap(working_array)
+      swapped_working_array = working_array.collect.with_index do |digit, index|
         swapper_map(index)[digit]
       end
+
+      return swapped_working_array
     end
 
     # Reverse swap
-    def unswap
-      @working_array = @working_array.collect.with_index do |digit, index|
+    def unswap(working_array)
+      unswapped_working_array = working_array.collect.with_index do |digit, index|
         swapper_map(index).rindex(digit)
       end
+
+      return unswapped_working_array
     end
 
     # Rearrange the order of each digit in a reversable way by using the 
     # sum of the digits (which doesn't change regardless of order)
     # as a key to record how they were scattered
-    def scatter
-      sum_of_digits = @working_array.inject(:+).to_i
-      @working_array = 10.times.collect do 
-        @working_array.rotate!(spin ^ sum_of_digits).pop
+    def scatter(working_array)
+      sum_of_digits = working_array.inject(:+).to_i
+      rotation = spin ^ sum_of_digits
+      scattered_working_array = 10.times.collect do
+        working_array.rotate!(rotation).pop
       end
+
+      return scattered_working_array
     end
 
     # Reverse the scatter
-    def unscatter
-      scattered_array = @working_array
+    def unscatter(working_array)
+      scattered_array = working_array
       sum_of_digits = scattered_array.inject(:+).to_i
-      @working_array = []
-      @working_array.tap do |unscatter| 
+      rotation = (sum_of_digits ^ spin) * -1
+      unscattered_working_array = []
+      unscattered_working_array.tap do |unscatter|
         10.times do
-          unscatter.push scattered_array.pop
-          unscatter.rotate! (sum_of_digits ^ spin) * -1
+          unscatter << scattered_array.pop
+          unscatter.rotate!(rotation)
         end
       end
+
+      return unscattered_working_array
     end
 
     # Add some spice so that different apps can have differently mapped hashes
     def spin
       @spin || 0
     end
+
+    # Right justifies the integer/string that's passed in creating an array of integers (representing digits in the final ID)
+    # Optimized for using arithmetic to left justify instead of string casting + rjusting
+    def build_working_array(original_integer)
+      zero_pad = [0,0,0,0,0,0,0,0,0,0]
+      rem = original_integer.to_i
+      denominator = 1_000_000_000
+      index = 0
+      while (rem > 0)
```

Comment:
On the last digit the denominator will be 1 and anything MOD 1 is always 0. I think the index < 10 check is clearer though.